knative-operator/pkg/common/telemetry/telemetry.go

```diff
@@ -0,0 +1,186 @@
+	switch t.name {
+	case "eventing":
+		sourcesG = serverlessTelemetryG.WithLabelValues("source")
+		pingSourceList := &eventingsourcesv1beta1.PingSourceList{}
+		if err := api.List(context.TODO(), pingSourceList); err == nil {
+			sourcesG.Set(float64(len(pingSourceList.Items)))
+			for _, obj := range pingSourceList.Items {
+				t.addToSnaphost(obj.GetObjectMeta())
+			}
+		}
+
+		sinkBindingList := &eventingsourcesv1beta1.SinkBindingList{}
+		if err := api.List(context.TODO(), sinkBindingList); err == nil {
+			sourcesG.Add(float64(len(sinkBindingList.Items)))
+			for _, obj := range sinkBindingList.Items {
+				t.addToSnaphost(obj.GetObjectMeta())
+			}
+		}
+
+		apiServerSourceList := &eventingsourcesv1beta1.ApiServerSourceList{}
+		if err := api.List(context.TODO(), apiServerSourceList); err == nil {
+			sourcesG.Add(float64(len(apiServerSourceList.Items)))
+			for _, obj := range pingSourceList.Items {
+				t.addToSnaphost(obj.GetObjectMeta())
+			}
+		}
+
+	case "knativeKafka":
+		knativeKafkaList := &kafkasourcev1beta1.KafkaSourceList{}
+		if err := api.List(context.TODO(), knativeKafkaList); err == nil {
+			sourcesG.Add(float64(len(knativeKafkaList.Items)))
+			for _, obj := range knativeKafkaList.Items {
+				t.addToSnaphost(obj.GetObjectMeta())
+			}
+		}
+
+	case "serving":
+		servicesG = serverlessTelemetryG.WithLabelValues("service")
+		serviceList := &servingv1.ServiceList{}
+		if err := api.List(context.TODO(), serviceList); err == nil {
+			servicesG.Set(float64(len(serviceList.Items)))
+			for _, obj := range serviceList.Items {
+				t.addToSnaphost(obj.GetObjectMeta())
+			}
+		}
+
+		revisionsG = serverlessTelemetryG.WithLabelValues("revision")
+		revisionList := &servingv1.RevisionList{}
+		if err := api.List(context.TODO(), revisionList); err != nil {
+			revisionsG.Set(float64(len(revisionList.Items)))
+			for _, obj := range revisionList.Items {
+				t.addToSnaphost(obj.GetObjectMeta())
+			}
+		}
+
+		routesG = serverlessTelemetryG.WithLabelValues("route")
+		routeList := &servingv1.RouteList{}
+		if err := api.List(context.TODO(), routeList); err == nil {
+			routesG.Set(float64(len(routeList.Items)))
+			for _, obj := range routeList.Items {
+				t.addToSnaphost(obj.GetObjectMeta())
+			}
+		}
+	}
+}
+
+func (t *Telemetry) addToSnaphost(obj metav1.Object) {
+	if obj == nil {
+		return
+	}
+	t.metricsSnapshot.Store(types.NamespacedName{
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	}, true)
+}
+
+func (t *Telemetry) deleteFromSnaphost(obj metav1.Object) {
+	if obj == nil {
+		return
+	}
+	t.metricsSnapshot.Delete(types.NamespacedName{
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	})
+}
+
+func (t *Telemetry) inSnapshot(obj metav1.Object) (ok bool) {
+	if obj == nil {
+		return false
+	}
+	_, ok = t.metricsSnapshot.Load(types.NamespacedName{
+		Namespace: obj.GetNamespace(),
+		Name:      obj.GetName(),
+	})
+	return
```

Comment:
please don't use anonymous returns. Adding `return ok` makes this a lot clearer imo.

Reply:
ok.